File: src/lib/wow/archive/client/raw-cache.ts
/**
 * Shared on-disk cache for raw (BLTE-decoded) WoW files, keyed by build + fileDataID.
 *
 * Layout: .cache/wow/data/<buildKey>/<fileDataID>
 *
 * Both wow-data-server and the converter resolve the same directory (relative to
 * the repo cwd), so when they run on the same host the converter reads files the
 * server already cached without any network hop.
 */
import { randomBytes } from 'crypto';
import fs from 'fs';
import {
  mkdir, readFile, rename, writeFile,
} from 'fs/promises';
import path from 'path';

import { constants } from '@/lib/wow/formats/constants';

const RAW_DATA_DIR = path.join(constants.DATA_PATH, 'data');

export function rawFileCachePath(buildKey: string, fileDataID: number): string {
  return path.join(RAW_DATA_DIR, buildKey, String(fileDataID));
}

/** Read a cached raw file, or null when absent. */
export async function readRawCachedFile(buildKey: string, fileDataID: number): Promise<Buffer | null> {
  try {
    const buf = await readFile(rawFileCachePath(buildKey, fileDataID));
    return buf.length > 0 ? buf : null;
  } catch {
    return null;
  }
}

export function rawCachedFileExistsSync(buildKey: string, fileDataID: number): boolean {
  return fs.existsSync(rawFileCachePath(buildKey, fileDataID));
}

/** Atomically persist a raw file (tmp + rename, safe under concurrent writers). */
export async function writeRawCachedFile(buildKey: string, fileDataID: number, data: Buffer): Promise<void> {
  const dest = rawFileCachePath(buildKey, fileDataID);
  await mkdir(path.dirname(dest), { recursive: true });
  const tmp = `${dest}.${randomBytes(6).toString('hex')}.tmp`;
  try {
    await writeFile(tmp, data);
    await rename(tmp, dest);
  } catch (e) {
    try { fs.unlinkSync(tmp); } catch { /* ignore */ }
    // A concurrent writer may have won the rename race; treat as success if so.
    if (!fs.existsSync(dest)) throw e;
  }
}

File: golang/internal/wow/archive/client/raw_cache.go
// Package client provides shared on-disk cache helpers for raw WoW files.
package client

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strconv"

	"github.com/pqhuy98/wow-converter/internal/wow/constants"
)

var rawDataDir = filepath.Join(constants.DataPath, "data")

// RawFileCachePath returns the cache path for a raw file.
func RawFileCachePath(buildKey string, fileDataID int) string {
	return filepath.Join(rawDataDir, buildKey, strconv.Itoa(fileDataID))
}

// ReadRawCachedFile reads a cached raw file, or nil when absent.
func ReadRawCachedFile(buildKey string, fileDataID int) ([]byte, error) {
	data, err := os.ReadFile(RawFileCachePath(buildKey, fileDataID))
	if err != nil {
		return nil, nil
	}
	if len(data) == 0 {
		return nil, nil
	}
	return data, nil
}

// RawCachedFileExistsSync reports whether a cached raw file exists.
func RawCachedFileExistsSync(buildKey string, fileDataID int) bool {
	_, err := os.Stat(RawFileCachePath(buildKey, fileDataID))
	return err == nil
}

// WriteRawCachedFile atomically persists a raw file.
func WriteRawCachedFile(buildKey string, fileDataID int, data []byte) error {
	dest := RawFileCachePath(buildKey, fileDataID)
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return err
	}
	tmpBytes := make([]byte, 6)
	if _, err := rand.Read(tmpBytes); err != nil {
		return err
	}
	tmp := fmt.Sprintf("%s.%s.tmp", dest, hex.EncodeToString(tmpBytes))
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	if err := os.Rename(tmp, dest); err != nil {
		_ = os.Remove(tmp)
		if _, statErr := os.Stat(dest); statErr != nil {
			return err
		}
	}
	return nil
}
